File: examples/phoenix/assets/js/hooks/line_chart.js
/**
 * D3 Line Chart hook — example component for the demo app.
 *
 * Multi-line chart with interactive points and tooltips. Pairs with
 * `D3ExDemoWeb.Components.Charts.LineChart`.
 *
 * Built on `createD3Hook` from the D3Ex library. The library itself ships only
 * the bridge primitives; this file is what an app developer would write.
 */

import { createD3Hook } from "../../../../priv/static/js/d3_hooks.js";

export const D3LineChart = {
  ...createD3Hook({
    onMount() {
      this.data = this.getData();
      this.initChart();
    },
    events: {
      set_data({ data })  { this.data = data;                    this.renderChart(); },
      append({ items })   { this.data = this.data.concat(items); this.renderChart(); },
      patch({ changes })  { this.applyPatch(changes);            this.renderChart(); },
      remove({ ids })     { this.applyRemove(ids);               this.renderChart(); },
    },
  }),

  applyPatch(changes) {
    const idKey = this.config.x_key;
    for (const { key, changes: itemChanges } of changes) {
      const item = this.data.find(d => d[idKey] === key);
      if (item) Object.assign(item, itemChanges);
    }
  },

  applyRemove(ids) {
    const idKey = this.config.x_key;
    const drop = new Set(ids);
    this.data = this.data.filter(d => !drop.has(d[idKey]));
  },

  initChart() {
    const d3 = window.d3;
    const { width, height, margin } = this.config;

    this.svg = d3.select(this.el).select('svg');
    this.g = this.svg.append('g')
      .attr('transform', `translate(${margin.left},${margin.top})`);

    this.innerWidth = width - margin.left - margin.right;
    this.innerHeight = height - margin.top - margin.bottom;

    this.xScale = d3.scaleLinear().range([0, this.innerWidth]);
    this.yScale = d3.scaleLinear().range([this.innerHeight, 0]);
    this.colorScale = d3.scaleOrdinal(d3[this.config.color_scheme] || d3.schemeCategory10);

    this.xAxis = this.g.append('g')
      .attr('class', 'x-axis')
      .attr('transform', `translate(0,${this.innerHeight})`);

    this.yAxis = this.g.append('g').attr('class', 'y-axis');

    if (this.config.show_grid) {
      this.xGrid = this.g.append('g')
        .attr('class', 'grid x-grid')
        .style('stroke', '#e0e0e0')
        .style('stroke-opacity', 0.3);

      this.yGrid = this.g.append('g')
        .attr('class', 'grid y-grid')
        .style('stroke', '#e0e0e0')
        .style('stroke-opacity', 0.3);
    }

    this.linesGroup = this.g.append('g').attr('class', 'lines');
    this.pointsGroup = this.g.append('g').attr('class', 'points');

    this.renderChart();
  },

  renderChart() {
    const d3 = window.d3;
    const { x_key, y_key, series_key, curve_type, show_points, show_area, animation_duration } = this.config;

    let seriesData;
    if (series_key) {
      const grouped = d3.group(this.data, d => d[series_key]);
      seriesData = Array.from(grouped, ([key, values]) => ({
        key,
        values: values.sort((a, b) => a[x_key] - b[x_key]),
      }));
    } else {
      seriesData = [{
        key: 'default',
        values: this.data.slice().sort((a, b) => a[x_key] - b[x_key]),
      }];
    }

    this.xScale.domain(d3.extent(this.data, d => d[x_key]));
    this.yScale.domain([0, d3.max(this.data, d => d[y_key])]);

    this.xAxis.transition().duration(animation_duration).call(d3.axisBottom(this.xScale));
    this.yAxis.transition().duration(animation_duration).call(d3.axisLeft(this.yScale));

    if (this.config.show_grid) {
      this.xGrid.transition().duration(animation_duration)
        .call(d3.axisBottom(this.xScale).tickSize(-this.innerHeight).tickFormat(''));

      this.yGrid.transition().duration(animation_duration)
        .call(d3.axisLeft(this.yScale).tickSize(-this.innerWidth).tickFormat(''));
    }

    const curveTypes = {
      'linear': d3.curveLinear,
      'monotone': d3.curveMonotoneX,
      'step': d3.curveStep,
    };

    const line = d3.line()
      .x(d => this.xScale(d[x_key]))
      .y(d => this.yScale(d[y_key]))
      .curve(curveTypes[curve_type] || d3.curveMonotoneX);

    let area;
    if (show_area) {
      area = d3.area()
        .x(d => this.xScale(d[x_key]))
        .y0(this.innerHeight)
        .y1(d => this.yScale(d[y_key]))
        .curve(curveTypes[curve_type] || d3.curveMonotoneX);

      const areas = this.linesGroup.selectAll('.area').data(seriesData, d => d.key);
      areas.exit().remove();

      const areasEnter = areas.enter()
        .append('path')
        .attr('class', 'area')
        .attr('fill-opacity', 0.2);

      areasEnter.merge(areas)
        .transition()
        .duration(animation_duration)
        .attr('d', d => area(d.values))
        .attr('fill', d => this.colorScale(d.key));
    }

    const lines = this.linesGroup.selectAll('.line').data(seriesData, d => d.key);
    lines.exit().remove();

    const linesEnter = lines.enter()
      .append('path')
      .attr('class', 'line')
      .attr('fill', 'none')
      .attr('stroke-width', 2);

    linesEnter.merge(lines)
      .transition()
      .duration(animation_duration)
      .attr('d', d => line(d.values))
      .attr('stroke', d => this.colorScale(d.key));

    if (show_points) {
      const points = this.pointsGroup
        .selectAll('circle')
        .data(this.data, (d, i) => `${d[x_key]}-${d[y_key]}-${i}`);

      points.exit().remove();

      const pointsEnter = points.enter()
        .append('circle')
        .attr('r', this.config.point_radius)
        .attr('stroke', '#fff')
        .attr('stroke-width', 1.5)
        .style('cursor', 'pointer');

      const pointsMerged = pointsEnter.merge(points);

      pointsMerged
        .on('click', (event, d) => this.sendEvent('on_point_click', d))
        .on('mouseover', (event) => {
          d3.select(event.currentTarget)
            .transition()
            .duration(150)
            .attr('r', this.config.point_radius * 1.5);
        })
        .on('mouseout', (event) => {
          d3.select(event.currentTarget)
            .transition()
            .duration(150)
            .attr('r', this.config.point_radius);
        });

      pointsMerged
        .transition()
        .duration(animation_duration)
        .attr('cx', d => this.xScale(d[x_key]))
        .attr('cy', d => this.yScale(d[y_key]))
        .attr('fill', d => series_key ? this.colorScale(d[series_key]) : this.colorScale('default'));
    }
  },
};

File: examples/phoenix/lib/d3_ex_demo_web/components/charts/line_chart.ex
defmodule D3ExDemoWeb.Components.Charts.LineChart do
  @moduledoc """
  Example multi-line chart built on `D3Ex.Component`. Lives in the demo app,
  not the library — D3Ex itself ships only the bridge primitives.

  Pairs with `assets/js/hooks/line_chart.js`.

  ## Example

      <.line_chart
        id="trends-chart"
        initial_data={@time_series_data}
        x_key={:date}
        y_key={:value}
        series_key={:metric}
        on_point_click="point_clicked"
        width={800}
        height={400}
      />

  `:initial_data` is consumed once at mount. For subsequent updates, use
  `D3Ex.Live.set_data/3`, `append/3`, `patch/3`, or `remove/3`.

  ## Options

  - `x_key`, `y_key` - Keys for the axes
  - `series_key` - Optional key for grouping multiple lines
  - `curve_type` - "linear" | "monotone" | "step" (default: "monotone")
  - `show_points`, `show_area`, `show_grid` - Visual toggles
  - `on_point_click`, `on_line_hover` - LiveView event names
  """

  use D3Ex.Component

  @impl true
  def default_config do
    %{
      width: 800,
      height: 400,
      margin: %{top: 20, right: 80, bottom: 40, left: 60},
      color_scheme: "schemeCategory10",
      curve_type: "monotone",
      show_points: true,
      show_area: false,
      show_grid: true,
      animation_duration: 750,
      point_radius: 4
    }
  end

  @impl true
  def prepare_assigns(assigns) do
    assigns
    |> Map.put_new(:initial_data, [])
    |> Map.put_new(:x_key, :x)
    |> Map.put_new(:y_key, :y)
    |> Map.put_new(:series_key, nil)
    |> Map.put_new(:x_label, nil)
    |> Map.put_new(:y_label, nil)
    |> Map.put_new(:on_point_click, nil)
    |> Map.put_new(:on_line_hover, nil)
  end

  @impl true
  def render(assigns) do
    ~H"""
    <div
      id={@id}
      phx-hook="D3LineChart"
      data-items={encode_data(@initial_data)}
      data-config={encode_config(Map.merge(@config, %{
        x_key: @x_key,
        y_key: @y_key,
        series_key: @series_key,
        x_label: @x_label,
        y_label: @y_label
      }))}
      data-events={encode_events(%{
        on_point_click: @on_point_click,
        on_line_hover: @on_line_hover
      })}
      phx-update="ignore"
      class="d3-line-chart"
      style={"width: #{@config.width}px; height: #{@config.height}px;"}
    >
      <svg width={@config.width} height={@config.height}></svg>
    </div>
    """
  end
end
